File: src/main/java/domain/TryCount.java
package domain;

import static message.ErrorMessage.TRY_COUNT_FORMAT_ERROR_MESSAGE;
import static message.ErrorMessage.TRY_COUNT_NEGATIVE_ERROR_MESSAGE;

public class TryCount {

    private final int value;

    private TryCount(int value) {
        validateNegative(value);
        this.value = value;
    }

    public static TryCount from(String value) {
        try {
            return new TryCount(Integer.parseInt(value));
        } catch (NumberFormatException e) {
            throw new IllegalArgumentException(TRY_COUNT_FORMAT_ERROR_MESSAGE.getValue());
        }
    }

    private void validateNegative(int value) {
        if (value < 0) {
            throw new IllegalArgumentException(TRY_COUNT_NEGATIVE_ERROR_MESSAGE.getValue());
        }
    }

    public TryCount decrease() {
        return new TryCount(value - 1);
    }

    public boolean isZero() {
        return value == 0;
    }
}

File: src/main/java/domain/TryCount.kt
package domain

import message.ErrorMessage.TRY_COUNT_FORMAT_ERROR_MESSAGE
import message.ErrorMessage.TRY_COUNT_NEGATIVE_ERROR_MESSAGE

class TryCount private constructor(value: Int) {
    private val value: Int

    init {
        validateNegative(value)
        this.value = value
    }

    private fun validateNegative(value: Int) {
        require(value >= 0) { TRY_COUNT_NEGATIVE_ERROR_MESSAGE.value }
    }

    fun decrease(): TryCount {
        return TryCount(value - 1)
    }

    val isZero: Boolean
        get() = value == 0

    companion object {
        @JvmStatic
        fun from(value: String): TryCount {
            return try {
                TryCount(value.toInt())
            } catch (e: NumberFormatException) {
                throw IllegalArgumentException(TRY_COUNT_FORMAT_ERROR_MESSAGE.value)
            }
        }
    }
}
